Fix: image size customisation controls not shown when theme doesn't define them but Woo core does

Konamiman opened the pull request:
**NOTE:** Closed because it diverged too much from master. Superseded by https://github.com/woocommerce/woocommerce/pull/28187.

### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

**NOTE:** merging https://github.com/woocommerce/woocommerce/pull/26731 is a prerequisite before this pull request can be merged. Also, base must be changed to `master` before merging.

### Changes proposed in this Pull Request:

In Customize » WooCommerce » Product Images, there should be Main Image Width options, for themes that do not declare image widths. However these controls don't show for the TwentyX themes, since the `includes/theme-support/class-wc-twenty-*.php` classes define them at initialization; the code that renders the controls (in the `WC_Shop_Customizer` class) is simply checking if the settings exist, and can't know who created them.

This pull request introduces the concept of _default_ theme support settings. The theme support classes define the image sizes as defaults, and the code that renders the controls do it only if the image sizes have been defined but not as defaults.

Also, for the cases where the user hasn't explicitly defined image and thumbnail sizes now the default values shown in the editor are the ones defined as theme support (default or not) instead of the hardcoded values of 600 and 300, respectively.

Closes #25900.

### How to test the changes in this Pull Request:

1. Set the Twenty-Twenty theme.
2. Go to Customize » WooCommerce » Product Images and verify that now you can change the image and thumbnail sizes.
3. Add theme support settings for image sizes on the Twenty-Twenty theme and verify that now you can't customize the image and thumbnail sizes. You can add these settings by adding the following code to the `twentytwenty_theme_support` method in `wp-content/themes/twentytwenty/functions.php`: `add_theme_support('woocommerce', array('thumbnail_image_width' => 450,'single_image_width' => 600));`
4. Undo the previous change and delete thumbnail image width size the by running this in the command line: `wp option delete woocommerce_thumbnail_image_width`. Check that the customizer UI now displays for this setting a value of 450, which is the value set in `WC_Twenty_Twenty::init`, instead of the hardcoded value of 300 (if you remove the code that sets the value in `WC_Twenty_Twenty::init` it will fallback to 300).

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Controls to change image and thumbnail sizes in theme customisation UI are now displayed even if there's a theme support class defining default values for those.

claudiosanches commented on includes/class-woocommerce.php:
```diff
@@ -8,6 +8,9 @@
 
 defined( 'ABSPATH' ) || exit;
 
+require_once __DIR__ . '/../src/Theming/ThemeSupport.php';
```
It will get already included by the autoloader.

Konamiman replied:
The problem is that I'm using the `ThemeSupport` class before plugins are loaded, and if I remove that line I get the following:

> Notice: Automattic\WooCommerce\Theming\ThemeSupport was called <strong>incorrectly</strong>. Not all plugins have loaded yet but we requested the class Automattic\WooCommerce\Theming\ThemeSupport Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version dev-fix/25900.) in /Users/konamiman/localsites/wp-54/app/public/wp-includes/functions.php on line 5167

ObliviousHarmony replied:
This is a limitation of the Jetpack Autoloader. There is a [PR being worked on](https://github.com/Automattic/jetpack/pull/15106) for a 2.0 of the autoloader that removes this limitation, but for the time being, this is probably the only way we can get around it. Admittedly it entirely defeats the point of the autoloader, but, we can't inject the dependencies otherwise.

claudiosanches commented on includes/wc-core-functions.php:
```diff
@@ -886,7 +858,7 @@ function wc_print_js() {
 		 * @param string $js JavaScript code.
 		 */
-		echo apply_filters( 'woocommerce_queued_js', $js ); // WPCS: XSS ok.
+		echo apply_filters( 'woocommerce_queued_js', $js ); //phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
```
Missing space after `//`.

claudiosanches commented on includes/wc-core-functions.php:
```diff
@@ -996,11 +969,13 @@ function flush_rewrite_rules_on_shop_page_save() {
 	}
 
-	$post_id      = intval( $_GET['post'] ); // WPCS: input var ok, CSRF ok.
+	//phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
Missing space after `//`.

claudiosanches replied:
And in several other places of this file too.

claudiosanches commented on src/Theming/ThemeSupport.php:
```diff
@@ -0,0 +1,123 @@
+	 *
+	 * @return array An array with all the theme support options defined for the 'woocommerce' feature, or false if nothing has been defined for that feature.
+	 */
+	private function get_all_woocommerce_theme_support_options() {
```
We could simply this method name, to just `get_all_theme_support_options`.

Konamiman replied:
I named the method like that because it's adding the options under the "woocommerce" feature. But yeah, maybe that's an implementation detail that doesn't need to be part of the method name.

claudiosanches reviewed (changes requested):
Great work!

There's a few minor issues to fix, mostly around coding standards, but in general everything looks great!

Also you should run `composer dump-autoload` to map all classes, so there's no need to include it manually in `includes/class-woocommerce.php`.

Konamiman commented on includes/class-woocommerce.php:
```diff
@@ -266,6 +269,7 @@ private function define( $name, $value ) {
 		}
 	}
 
+	//phpcs:disable Generic.Commenting.Todo.CommentFound, WordPress.Security.ValidatedSanitizedInput
```
Ok, I thought that warnings would prevent me from committing.

Konamiman commented on includes/wc-core-functions.php:
```diff
@@ -731,40 +732,11 @@ function wc_mail( $to, $subject, $message, $headers = "Content-Type: text/html\r
  * @param  mixed  $default Optional value to return if the theme does not declare support for a prop.
  * @return mixed  Value of prop(s).
+ *
+ * @deprecated 4.3.0 Use WooCommerce::theme_support()->get_woocommerce_theme_support instead.
```
Maybe deprecating isn't the best option, but I think we should somehow encouraging using the new class instead of the function for new code, as code will be easier to test that way.

ObliviousHarmony commented on src/Theming/ThemeSupport.php:
```diff
@@ -0,0 +1,123 @@
+	 *
+	 * @param array $options The options to be added.
+	 */
+	public function add_theme_support_options( $options ) {
```
How would you feel about removing the `theme_support` from every method in this class? It seems redundant that the `ThemeSupport` class has a `add_theme_support_options` method instead of just an `add_options` method.

Konamiman replied:
Yep, that makes sense. I read too much non-object-oriented code already and it's spoiling my naming abilities. 😅

ObliviousHarmony commented on src/Utils/ArrayUtils.php:
```diff
@@ -0,0 +1,60 @@
+			throw new \Exception( StringUtils::get_class_short_name( __CLASS__ ) . '::' . __METHOD__ . ': $array must be an array.' );
+		}
+
+		$key_stack = explode( '::', $key );
```
Dot notation seems to be a more common pattern for these kinds of utilities in the PHP ecosystem, what do you think?

Konamiman replied:
The thing is that this is being used by the replacement of the `wc_get_theme_support` method, and we need to keep this separator for compatibility. But I could make the separator an optional parameter of the method, defaulting to `.`, and use `::` when invoked from the theme support class.

ObliviousHarmony replied:
It looks like `::` is a pretty common delimiter in the WordPress ecosystem so let's just leave it as it is I think!

ObliviousHarmony commented on src/Utils/ArrayUtils.php:
```diff
@@ -0,0 +1,60 @@
+	 */
+	public static function get_nested_value( $array, $key, $default = null ) {
+		if ( ! is_array( $array ) ) {
+			throw new \Exception( StringUtils::get_class_short_name( __CLASS__ ) . '::' . __METHOD__ . ': $array must be an array.' );
```
How about we replace this with an `array` type on `$array`?

Konamiman replied:
Yep, makes perfect sense.

ObliviousHarmony commented on includes/class-woocommerce.php:
```diff
@@ -844,6 +854,15 @@ public function mailer() {
+	 *
+	 * @return ThemeSupport
+	 */
+	public static function theme_support() {
```
As we discussed in Slack, I think we might benefit from introducing a dependency container now, as opposed to further extending this global.

Konamiman replied:
On it!

ObliviousHarmony reviewed (changes requested):
I am in LOVE with this PR, really great work!

Other than the comments I've left, the only change I'd want to see is type declarations! Since our minimum version is 7.0 we are now able to use function type declarations as well as return types! I feel like any new code written in the namespace should fully utilize the feature set available to our minimum version.

Konamiman committed:
First steps towards introducing a dependency injection framework.

- Add PHP League's Container package via Composer.
- Add an ObjectContainer class that encapsulates all the configuration
  and insulates the codebase from the concrete DI engine used.
- Add an improved ReflectionContainer class that will allow to
  register individual classes as singletons while autowiring.
- Use ObjectContainer to resolve the WooCommerce class, everything
  instantiated with "new" inside it, and all singletons that are
  usually obtained via WC() function.
- Introduce the CustomerProvider class.
- Introduce a service provider to resolve WC_Queue_Interface,
  this replaces the WC_Queue class.
- Mark as obsolete all the replaced "instance()" methods,
  and the entire WC_Queue class.

Konamiman committed:
Move PSR4 autoloading from woocommerce.php to the Autoloader class

Konamiman committed:
Undo the registration of old singletons in the DI container.

The container will in principle be used only to register classes
in the `src` directory.

Also, CustomerProvider class removed.

Konamiman committed:
Changes in the dependency injection container:

- Renamed from `ObjectContainer` to `Container`.
- It now inherits from PHP League's `Container`.
- It has now a `defineAsSharedAutowired` method.
- Initialization moved to the `WooCommerce::init_container` method.
- The static method for object resolution is now `WooCommerce::get_instance_of`.

Konamiman committed:
Revert a couple of files to its original content after a messed rebase

Konamiman committed:
Changes in the overall organization of the dependency injection:

- The `Container` class now implements `Psr\Container\ContainerInterface`
  (and registers itself as such), holding a private instance of the
  real container. This way it's a read-only container from the point
  of view of plugins (which should use their own containers, but
  can still use this to get WooCommerce classes).
- All registrations are now done in the `Container` constructor via
  service providers.
- The container instance is now held in a global variable, set in
  `woocommerce.php`
- Added the `wc_get_container` function for old code.
- Added the `AbstractServiceProvider` class, which inherits with the
  corresponding League's class and adds some utility methods,
  most notably `add/shareWithAutoArguments`.
- Added the `ActionsProxy` and `LegacyProxy` classes, they are
  registered via a dedicated service provider.
- `WC_Queue_Interface` is no longer resolvable via the container
  (which is for classes inside `src` only).
- All the method names in the new classes have the format `fooBarFizz`
  to be PSR4 compliant, so the MethodNameInvalid error has been
  disabled in phpcs.xml for the `src` directory.
- Introduced the `@public` annotation for public API classes
  (classes that plugins can use and whose backwards compatibility
  we guarantee), applied to `ActionsProxy` and to `LegacyProxy` for now.
- Removed the hack for the autoloader as now it doesn't work anyway.
  For the changes in this branch to work, now WP_DEBUG must be false.

Konamiman committed:
Some changes on the dependency injection engine based on feedback:

- `wc_get_container` declares a return type now.
- The autoloader now runs before `autoload_packages.php` is included.
- The autloader now excludes the namespaces for Admin, Blocs, Rest API.
- `private $container` in `Container` is now declared as being
  of type `\League\Container\Container`.
- The underlying container is registered with `share` instead of `add`.
- All methods in `AbstractServiceProvider` now allow passing a concrete.
- `AbstractServiceProvider::addWithAutoArguments` now checks if an
  argument has a default value, and registers the argument so that
  that value is supplied when constructing the object.
- `Proxies` class renamed to `ProxiesServiceProvider`.
- The methods to call code in `LegacyProxy` now use `call_user_func_array`.

Konamiman committed:
`DependencyManagement` and `Proxies` directories moved.

They have been moved out of the `Tools` directory to the root
`src` directory, `Tools` doesn't exist anymore.

Konamiman committed:
Added dependency injection infrastructure for unit tests:

- Added a new class `ExtendedContainer` that extends League's container.
  - `add` modified to reject classes not in the root Woo namespace.
  - Has two new methods, `replace` and `reset_resolved`.
  - It's used as the underlying container instead of League's one
    in `Container`, but the new methods are not exposed.
- At unit test bootstrap time the globally registered container is
  replaced with the extended one that `Container` stores
  (grabbed from private property using reflection).
- A new `MockableLegacyProxy` is added. It inherits from `LegacyProxy`
  and allows to mock functions, static methods and legacy classes.
  - The registeed `LegacyProxy` is replaced with the mockable version
    during unit test bootstrap.
- A PHPUnit hook is added to reset the mockable proxy to its initial
  state (so that nothing is mocked) before each test.
- `WC_Unit_Test_Case` gets helper methods to mock functions, static
  methods and classes without having to retrieve the proxy class.

Konamiman committed:
Fix code sniffer errors in wc-core-functions.php

Konamiman committed:
Some small imrpovements in the dependency injection framework:

- camelCase methods changed to snake_case for consistency with WP.
- Added a check in `ExtendedContainer::get` that throws an informative
  exception if a non-namespaced class name is passed.
- `container->reset_resolved()` is called during unit testing bootstrap.
- Added some utility methods in `WC_Unit_Test_Case`.

Konamiman committed:
Fix #25900 - image size customisation controls not shown

- Added the `ThemeSupport class`, with methods to add and get
  theme support options.
  - It also has a new `add_default_options` method that adds the
    options under a `_defaults` key.
- The `WC_Twenty_*` classes now use `ThemeSupport` instead of
  the `add_theme_support` function to define image and thumbnail sizes.
  - The values are defined as default options.
- The `WC_Shop_Customizer` class now uses `ThemeSupport` instead of
  `wc_get_theme_support` to check if image and thumbnail sizes UI
  should be rendered.
  - The check is made excluding default values.

With these changes the UI to change the image and thumbnail sizes
is hidden only if the options are added as non-defaults elsewhere.

Additional changes:

- The code of the `wc_get_theme_support` function is replaced with
  a simple call to `get_option` in `ThemeSupport`.
- Added the utility classes `StringUtils` and `ArrayUtils`.

Konamiman committed:
Add tests for `AbstractServiceProvider` and `ExtendedContainer`.

Also:

- Make the methods in `AbstractServiceProvider` protected.
- Add an autoloader for files in the `tests/php/src` directory.
- Fix a bug in the provisional (?) autoloader.

Konamiman committed:
Added unit tests for `LegacyProxy` and `MockableLegacyProxy`.

Also:

- Added sutoloader for the `tests/tools` directory.

Konamiman committed:
Add `@public` to `FunctionsMockerHack`.

Also to a couple of methods related to the code hacker.

Konamiman committed:
Rename namespace "Theming" to "ThemeManagement"

Konamiman commented:
Closed because it diverged too much from master. Superseded by https://github.com/woocommerce/woocommerce/pull/28187.
